src/names.c

```diff
@@ -441,6 +442,88 @@ SEXP outer_names(SEXP names, SEXP outer, R_len_t n) {
   }
 }
 
+// [[ register() ]]
+SEXP vctrs_apply_name_spec(SEXP name_spec, SEXP outer, SEXP inner, SEXP n) {
+  return apply_name_spec(name_spec, r_chr_get(outer, 0), inner, r_int_get(n, 0));
+}
+
+static SEXP glue_as_name_spec(SEXP spec);
+
+// [[ include("utils.h") ]]
+SEXP apply_name_spec(SEXP name_spec, SEXP outer, SEXP inner, R_len_t n) {
+  if (outer == R_NilValue) {
+    return inner;
+  }
+  if (TYPEOF(outer) != CHARSXP) {
+    Rf_error("Internal error: `outer` must be a scalar string.");
+  }
+
+  if (outer == strings_empty || outer == NA_STRING) {
+    return inner;
+  }
+
+  if (r_is_empty_names(inner)) {
+    if (n == 1) {
+      return r_str_as_character(outer);
+    }
+    inner = PROTECT(r_seq(1, n + 1));
+  } else {
+    inner = PROTECT(inner);
+  }
+
+  switch (TYPEOF(name_spec)) {
```

Comment:
Would it make more sense to do this in R?

Reply:
The C switch doesn't seem bad.

Reply:
Plus the 3 new helper functions 😉

Reply:
oh I think I'll export an `rlang_as_function()` C callable. Here I'm also experimenting with passing an `arg` param, maybe that should become standard in all dev-oriented tools.